File: server/utils/cdn/tarball.ts
import { HTTPError } from "nitro/h3";

import { cacheStorage } from "../storage";
import { MAX_UNPACKED_SIZE, SKIP_TTL_MS, TARBALL_DOWNLOAD_TIMEOUT } from "./constants";
import { calculateIntegrity } from "./integrity";
import type { CdnFile } from "./types";

// Concurrency guard: prevent duplicate tarball downloads for the same cacheBase
const pendingTarballs = new Set<string>();

/**
 * Options for tarball root directory detection.
 */
export interface RootDirOptions {
  /**
   * If true, skip entries starting with 'pax_global_header' when finding
   * the root directory. Needed for GitHub tarballs which include
   * Pax header entries.
   * Default: false
   */
  skipPaxHeaders?: boolean;

  /**
   * Fallback directory name if root cannot be detected from tarball entries.
   * Default: "package"
   */
  fallbackName?: string;
}

/**
 * Detect the root directory name inside a tarball's file map.
 */
export function detectRootDir(filesMap: Map<string, File>, options?: RootDirOptions): string {
  const keys = [...filesMap.keys()];
  let firstPath: string | undefined;

  if (options?.skipPaxHeaders) {
    firstPath = keys.find((p) => p.includes("/") && !p.startsWith("pax_global_header"));
  } else {
    firstPath = keys[0];
  }

  const fallback = options?.fallbackName ?? "package";
  return firstPath?.split("/")[0] || fallback;
}

/**
 * Download a tarball with timeout and return its bytes.
 * Throws HTTPError on timeout or non-OK response.
 */
export async function downloadTarball(
  tarballUrl: string,
  errorContext = "package",
): Promise<Uint8Array> {
  let tarballRes: Response;
  try {
    tarballRes = await fetch(tarballUrl, { signal: AbortSignal.timeout(TARBALL_DOWNLOAD_TIMEOUT) });
  } catch (error) {
    if ((error as Error).name === "AbortError") {
      throw new HTTPError({
        status: 504,
        statusText: `Tarball download timed out, ${errorContext} may be too large`,
      });
    }
    throw new HTTPError({ status: 502, statusText: "Failed to download tarball" });
  }

  if (!tarballRes.ok) {
    throw new HTTPError({ status: 502, statusText: "Failed to download tarball" });
  }

  return tarballRes.bytes();
}

/**
 * Try fetching a URL with timeout and return bytes if successful.
 * Returns undefined on any failure (network error, non-OK status, timeout).
 */
async function tryFetchWithTimeout(url: string): Promise<Uint8Array | undefined> {
  try {
    const res = await fetch(url, { signal: AbortSignal.timeout(TARBALL_DOWNLOAD_TIMEOUT) });
    if (res.ok) {
      return res.bytes();
    }
  } catch {
    // Ignore network errors, timeouts, and non-OK responses
  }
  return undefined;
}

/**
 * Extract a single file from a tarball (cache-first, download on miss).
 *
 * If `directUrl` is provided, it will be tried before downloading the tarball.
 * Fallback chain for GitHub raw URLs: raw → raw/master → jsdelivr CDN → tarball.
 */
export async function extractFileFromTarball(
  tarballUrl: string,
  filepath: string,
  cacheKey: string,
  rootDirOptions?: RootDirOptions,
  directUrl?: string,
): Promise<Uint8Array> {
  const storage = cacheStorage;

  // Try cache first
  const cached = await storage.getItemRaw(cacheKey);
  if (cached) {
    return new Uint8Array(cached);
  }

  // Try direct URL and fallbacks
  if (directUrl) {
    // 1. Try raw.githubusercontent.com
    const data = await tryFetchWithTimeout(directUrl);
    if (data) {
      await storage.setItemRaw(cacheKey, data);
      return data;
    }

    // 2. If default main was used, try master and jsdelivr CDN
    if (directUrl.includes("/main/")) {
      const masterUrl = directUrl.replace(/\/main\//, "/master/");
      const masterData = await tryFetchWithTimeout(masterUrl);
      if (masterData) {
        await storage.setItemRaw(cacheKey, masterData);
        return masterData;
      }

      // jsdelivr CDN without version (resolves default branch automatically)
      const jsdelivrBase = directUrl.replace(
        "https://raw.githubusercontent.com/",
        "https://cdn.jsdelivr.net/gh/",
      );
      const urlObj = new URL(jsdelivrBase);
      const pathParts = urlObj.pathname.split("/");
      if (pathParts.length >= 6) {
        pathParts.splice(4, 1);
      }
      urlObj.pathname = pathParts.join("/");
      const jsdelivrData = await tryFetchWithTimeout(urlObj.toString());
      if (jsdelivrData) {
        await storage.setItemRaw(cacheKey, jsdelivrData);
        return jsdelivrData;
      }
    }
  }

  // Download tarball and extract the target file
  const tarballData = await downloadTarball(tarballUrl);
  const archive = new Bun.Archive(tarballData);

  const rootDir = await detectRootDirFromArchive(archive, rootDirOptions);
  const globPattern = `${rootDir}/${filepath}`;
  const matchedFiles = await archive.files(globPattern);
  const targetFile = matchedFiles.get(globPattern);

  if (!targetFile) {
    throw new HTTPError({
      status: 404,
      statusText: `File not found: ${filepath}`,
    });
  }

  return new Uint8Array(await targetFile.bytes());
}

/**
 * Detect root directory from a Bun.Archive without listing all files.
 */
async function detectRootDirFromArchive(
  archive: Bun.Archive,
  options?: RootDirOptions,
): Promise<string> {
  const filesMap = await archive.files();
  return detectRootDir(filesMap, options);
}

/**
 * Check if a package version's file list is already cached.
 */
export async function isPackageCached(
  cacheBase: string,
  isCacheableVersion: boolean,
): Promise<boolean> {
  if (!isCacheableVersion) return false;
  const storage = cacheStorage;
  const meta = await storage.getMeta(cacheBase);
  return !!meta?.files;
}

/**
 * Download and cache all files from a tarball.
 * Intended for background execution (via event.waitUntil or after response).
 */
export async function cachePackageFromTarball(
  tarballUrl: string,
  cacheBase: string,
  rootDirOptions?: RootDirOptions,
  logLabel?: string,
): Promise<void> {
  const storage = cacheStorage;

  // Check if already cached with file list, or still within skip TTL
  const cachedMeta = await storage.getMeta(cacheBase);
  if (cachedMeta?.files) return;
  if (cachedMeta?.skippedAt && Date.now() - Number(cachedMeta.skippedAt) < SKIP_TTL_MS) return;

  // Skip if another call is already downloading this tarball
  if (pendingTarballs.has(cacheBase)) return;
  pendingTarballs.add(cacheBase);

  try {
    // Download and extract tarball
    const tarballRes = await fetch(tarballUrl, {
      signal: AbortSignal.timeout(TARBALL_DOWNLOAD_TIMEOUT),
    });
    if (!tarballRes.ok) {
      console.error(`Failed to download tarball for ${logLabel || cacheBase}`);
      await storage.setMeta(cacheBase, { skippedAt: Date.now() });
      return;
    }

    const tarballData = await tarballRes.bytes();
    const archive = new Bun.Archive(tarballData);
    const filesMap = await archive.files();

    const rootDir = detectRootDir(filesMap, rootDirOptions);
    const rootPath = `${rootDir}/`;

    // Build file list metadata
    const fileList: Array<CdnFile> = [];
    for (const [path, file] of filesMap) {
      fileList.push({
        name: path.slice(rootPath.length),
        size: file.size,
      });
    }

    // Skip packages that exceed size limit
    const totalSize = fileList.reduce((sum, f) => sum + f.size, 0);
    if (totalSize > MAX_UNPACKED_SIZE) {
      console.warn(
        `Skipping ${logLabel || cacheBase}: unpacked size ${(totalSize / 1024 / 1024).toFixed(1)} MB exceeds ${MAX_UNPACKED_SIZE / 1024 / 1024} MB limit`,
      );
      await storage.setMeta(cacheBase, { skippedAt: Date.now() });
      return;
    }

    // Concurrent caching using Promise.allSettled
    const cachePromises = Array.from(filesMap.entries()).map(async ([path, file]) => {
      const relativePath = path.slice(rootPath.length);
      const cacheKey = `${cacheBase}/${relativePath}`;

      try {
        const exists = await storage.getItemRaw(cacheKey);
        if (exists) return;

        const fileData = await file.bytes();
        await storage.setItemRaw(cacheKey, fileData);

        // Calculate SHA-256 integrity for SRI
        const integrity = calculateIntegrity(fileData);
        const fileItem = fileList.find((f) => f.name === relativePath);
        if (fileItem) {
          fileItem.integrity = integrity;
        }
      } catch (error) {
        console.error(`Failed to cache file ${relativePath}:`, error);
      }
    });

    await Promise.allSettled(cachePromises);

    // Store file list in meta
    await storage.setMeta(cacheBase, { files: fileList });
  } catch (error) {
    console.error(`Background cache failed for ${logLabel || cacheBase}:`, error);
  } finally {
    pendingTarballs.delete(cacheBase);
  }
}

File: src/cdn/utils/tarball.rs
use anyhow::Result;
use flate2::read::GzDecoder;
use std::collections::{HashMap, HashSet};
use std::io::Read;
use std::sync::{LazyLock, Mutex};
use std::time::Duration;
use tar::Archive;
use tracing::{error, warn};

use super::constants::*;
use super::integrity::calculate_integrity;
use crate::storage::{CacheMeta, CdnFileMeta, SharedStorage};

/// Deduplication set for concurrent background-cache jobs (mirrors pendingTarballs).
static PENDING: LazyLock<Mutex<HashSet<String>>> = LazyLock::new(|| Mutex::new(HashSet::new()));

/// RAII guard that removes the cache_base from PENDING on drop, ensuring cleanup
/// across all return paths (including early returns and panics).
struct PendingGuard {
    key: String,
}

impl Drop for PendingGuard {
    fn drop(&mut self) {
        PENDING.lock().unwrap().remove(&self.key);
    }
}

pub struct TarEntry {
    pub name: String,
    pub data: Vec<u8>,
}

pub fn extract_tgz(data: &[u8]) -> Result<Vec<TarEntry>> {
    let decoder = GzDecoder::new(data);
    let mut archive = Archive::new(decoder);
    let mut entries = Vec::new();

    for entry in archive.entries()? {
        let mut entry = entry?;
        // Skip directory entries: they hold no data. On filesystem backends a
        // cached "dist"/"src" empty file shadows its child paths (a "dist" file
        // blocks writing "dist/d3-selection.js"), silently dropping every file
        // under that directory. Only regular files are cached.
        if entry.header().entry_type().is_dir() {
            continue;
        }
        let path = entry.path()?.to_string_lossy().to_string();
        let size = entry.size();
        let mut buf = Vec::with_capacity(size as usize);
        entry.read_to_end(&mut buf)?;
        entries.push(TarEntry {
            name: path,
            data: buf,
        });
    }

    Ok(entries)
}

pub fn extract_file_from_tgz(data: &[u8], filepath: &str) -> Option<Vec<u8>> {
    // Determine the root dir once with a dedicated scan, then match (mirrors tar.ts:
    // detectRootDirFromBytes first, then extractFileFromTgz matches against it).
    let root = detect_root_dir(data);
    let full_path = format!("{root}/{filepath}");

    let decoder = GzDecoder::new(data);
    let mut archive = Archive::new(decoder);

    for entry in archive.entries().ok()? {
        let mut entry = entry.ok()?;
        let path = entry.path().ok()?;
        if path.to_string_lossy() == full_path {
            let mut buf = Vec::new();
            entry.read_to_end(&mut buf).ok()?;
            return Some(buf);
        }
    }

    None
}

pub fn detect_root_dir(data: &[u8]) -> String {
    let decoder = GzDecoder::new(data);
    let mut archive = Archive::new(decoder);

    for entry in archive.entries().into_iter().flatten() {
        if let Ok(entry) = entry
            && let Ok(path) = entry.path()
        {
            let path_str = path.to_string_lossy();
            if !path_str.starts_with("pax_global_header") {
                return path_str.split('/').next().unwrap_or("package").to_string();
            }
        }
    }

    "package".to_string()
}

pub async fn download_tarball(url: &str) -> Result<Vec<u8>> {
    // Cap concurrent outbound fetches so a burst of cache misses doesn't trip
    // npm's per-IP rate limit (429 / IP block).
    let _permit = super::concurrency::DOWNLOAD_SEMAPHORE
        .acquire()
        .await
        .unwrap();

    let mut resp = crate::http::HTTP_CLIENT
        .get(url)
        .timeout(Duration::from_secs(CDN_FETCH_TIMEOUT_SECS))
        .send()
        .await
        .map_err(|e| {
        if e.is_timeout() {
            anyhow::anyhow!("Tarball download timed out")
        } else {
            anyhow::anyhow!("Failed to download tarball: {e}")
        }
    })?;

    if !resp.status().is_success() {
        anyhow::bail!("Failed to download tarball: {}", resp.status());
    }

    // Reject before reading the body when the server declares an oversized
    // Content-Length, then stream with a hard cap so a missing/lying header
    // (or a huge gh repo tarball) can't exhaust memory or bandwidth. The
    // post-extract CDN_MAX_PACKAGE_SIZE check in cache_package_from_tarball
    // still guards the *unpacked* size (tarballs compress).
    if let Some(len) = resp.content_length()
        && len > CDN_MAX_PACKAGE_SIZE
    {
        anyhow::bail!(
            "Tarball exceeds {} byte limit (declared {len})",
            CDN_MAX_PACKAGE_SIZE
        );
    }
    let mut body = Vec::new();
    while let Some(chunk) = resp.chunk().await? {
        if body.len() + chunk.len() > CDN_MAX_PACKAGE_SIZE as usize {
            anyhow::bail!(
                "Tarball exceeded {} byte limit while streaming",
                CDN_MAX_PACKAGE_SIZE
            );
        }
        body.extend_from_slice(&chunk);
    }
    Ok(body)
}

pub async fn extract_file_from_tarball(
    storage: &SharedStorage,
    tarball_url: &str,
    filepath: &str,
    cache_key: &str,
    direct_url: Option<&str>,
    warm: Option<(&str, &str)>,
) -> Result<Vec<u8>> {
    if let Some(cached) = storage.get_raw(cache_key).await {
        return Ok(cached);
    }

    // Single-flight: dedup concurrent cache-miss for the same key so only the
    // leader downloads; followers wait, then re-read storage.
    let storage_for_fn = storage.clone();
    let warm = warm.map(|(base, label)| (base.to_string(), label.to_string()));
    super::singleflight::run_once(cache_key, || {
        let storage = storage_for_fn.clone();
        let tarball_url = tarball_url.to_string();
        let filepath = filepath.to_string();
        let direct_url = direct_url.map(|u| u.to_string());
        let cache_key = cache_key.to_string();
        let warm = warm.clone();
        async move {
            // Re-check: another leader may have just cached it.
            if storage.get_raw(&cache_key).await.is_some() {
                return;
            }
            if let Some(url) = direct_url.as_deref() {
                if let Some(data) = try_fetch(url).await {
                    storage.set_raw(&cache_key, &data).await;
                    return;
                }
                if url.contains("/main/") {
                    let master_url = url.replace("/main/", "/master/");
                    if let Some(data) = try_fetch(&master_url).await {
                        storage.set_raw(&cache_key, &data).await;
                        return;
                    }
                    if let Ok(mut url_obj) = url::Url::parse(&url.replace(
                        "https://raw.githubusercontent.com/",
                        "https://cdn.jsdelivr.net/gh/",
                    )) {
                        let mut parts: Vec<&str> = url_obj.path().split('/').collect();
                        if parts.len() >= 6 {
                            parts.remove(4);
                        }
                        url_obj.set_path(&parts.join("/"));
                        if let Some(data) = try_fetch(url_obj.as_str()).await {
                            storage.set_raw(&cache_key, &data).await;
                            return;
                        }
                    }
                }
            }
            if let Ok(tarball) = download_tarball(&tarball_url).await {
                if let Some(data) = extract_file_from_tgz(&tarball, &filepath) {
                    storage.set_raw(&cache_key, &data).await;
                }
                // Warm the full package in the background reusing these bytes, so a
                // follow-up request for another file (or a directory listing) is served
                // from cache without re-downloading the tarball. Only the download path
                // has bytes to reuse; the direct_url fast path above never fetched the
                // tarball, so gh (which uses direct_url) warms via its own spawn.
                if let Some((base, label)) = warm {
                    let storage = storage.clone();
                    tokio::spawn(async move {
                        let _ = cache_package_from_bytes(&storage, tarball, &base, &label).await;
                    });
                }
            }
        }
    })
    .await;

    storage
        .get_raw(cache_key)
        .await
        .ok_or_else(|| anyhow::anyhow!("File not found: {filepath}"))
}

pub async fn is_package_cached(storage: &SharedStorage, cache_base: &str, cacheable: bool) -> bool {
    if !cacheable {
        return false;
    }
    storage
        .get_meta(cache_base)
        .await
        .and_then(|m| m.files)
        .is_some()
}

/// Skip if already cached (files present) or recently skipped; otherwise claim the
/// PENDING slot to dedup concurrent cache jobs. Returns a guard whose drop releases
/// the slot, or None when nothing should be done.
async fn try_acquire_cache_slot(
    storage: &SharedStorage,
    cache_base: &str,
) -> Option<PendingGuard> {
    if let Some(meta) = storage.get_meta(cache_base).await {
        if meta.files.is_some() {
            return None;
        }
        if let Some(skipped) = meta.skipped_at {
            let now = now_millis();
            if now - skipped < CDN_SKIP_TTL_MS {
                return None;
            }
        }
    }

    let mut set = PENDING.lock().unwrap();
    if set.contains(cache_base) {
        return None;
    }
    set.insert(cache_base.to_string());
    Some(PendingGuard {
        key: cache_base.to_string(),
    })
}

/// Extract every file from an already-obtained tarball into storage and write the
/// file-list meta. Shared by the download entry point and the byte-reuse warm path.
/// The caller must hold the PENDING slot (via `try_acquire_cache_slot`) so concurrent
/// jobs for the same package don't duplicate the extract.
async fn cache_package_entries(
    storage: &SharedStorage,
    tarball_data: &[u8],
    cache_base: &str,
    log_label: &str,
) -> Result<()> {
    let root_dir = detect_root_dir(tarball_data);
    let root_path = format!("{root_dir}/");
    let entries = extract_tgz(tarball_data)?;

    let filtered: Vec<&TarEntry> = entries
        .iter()
        .filter(|e| e.name.starts_with(&root_path))
        .collect();

    let mut file_list: Vec<CdnFileMeta> = filtered
        .iter()
        .map(|e| CdnFileMeta {
            name: e.name[root_path.len()..].to_string(),
            size: e.data.len() as u64,
            integrity: None,
        })
        .collect();

    let total_size: u64 = file_list.iter().map(|f| f.size).sum();
    if total_size > CDN_MAX_PACKAGE_SIZE {
        warn!(
            "Skipping {log_label}: unpacked size {} MB exceeds {} MB limit",
            total_size / 1024 / 1024,
            CDN_MAX_PACKAGE_SIZE / 1024 / 1024
        );
        storage
            .set_meta(
                cache_base,
                &CacheMeta {
                    skipped_at: Some(now_millis()),
                    ..Default::default()
                },
            )
            .await;
        return Ok(());
    }

    let mut file_index: HashMap<String, usize> = HashMap::new();
    for (i, f) in file_list.iter().enumerate() {
        file_index.insert(f.name.clone(), i);
    }

    for entry in &filtered {
        let relative = &entry.name[root_path.len()..];
        let key = format!("{cache_base}/{relative}");

        if storage.get_raw(&key).await.is_some() {
            continue;
        }

        storage.set_raw(&key, &entry.data).await;

        let integrity = calculate_integrity(&entry.data);
        if let Some(&idx) = file_index.get(relative) {
            file_list[idx].integrity = Some(integrity);
        }
    }

    storage
        .set_meta(
            cache_base,
            &CacheMeta {
                files: Some(file_list),
                ..Default::default()
            },
        )
        .await;

    Ok(())
}

pub async fn cache_package_from_tarball(
    storage: &SharedStorage,
    tarball_url: &str,
    cache_base: &str,
    log_label: &str,
) -> Result<()> {
    let _guard = match try_acquire_cache_slot(storage, cache_base).await {
        Some(g) => g,
        None => return Ok(()),
    };

    let tarball_data = match download_tarball(tarball_url).await {
        Ok(data) => data,
        Err(e) => {
            error!("Failed to download tarball for {log_label}: {e}");
            storage
                .set_meta(
                    cache_base,
                    &CacheMeta {
                        skipped_at: Some(now_millis()),
                        ..Default::default()
                    },
                )
                .await;
            return Ok(());
        }
    };

    cache_package_entries(storage, &tarball_data, cache_base, log_label).await
}

/// Cache a full package from tarball bytes the caller already downloaded (e.g. the
/// foreground sub-path request that extracted one file). Avoids a second tarball
/// download for the background warm path. Same skip/dedup semantics as
/// `cache_package_from_tarball`.
pub async fn cache_package_from_bytes(
    storage: &SharedStorage,
    tarball_data: Vec<u8>,
    cache_base: &str,
    log_label: &str,
) -> Result<()> {
    let _guard = match try_acquire_cache_slot(storage, cache_base).await {
        Some(g) => g,
        None => return Ok(()),
    };

    cache_package_entries(storage, &tarball_data, cache_base, log_label).await
}

pub async fn try_fetch(url: &str) -> Option<Vec<u8>> {
    let _permit = super::concurrency::DOWNLOAD_SEMAPHORE
        .acquire()
        .await
        .ok()?;
    let mut resp = crate::http::HTTP_CLIENT
        .get(url)
        .timeout(Duration::from_secs(CDN_FETCH_TIMEOUT_SECS))
        .send()
        .await
        .ok()?;
    if !resp.status().is_success() {
        return None;
    }
    // Same streaming size cap as download_tarball: wp zips and gh raw files
    // flow through here and either can be oversized.
    if let Some(len) = resp.content_length()
        && len > CDN_MAX_PACKAGE_SIZE
    {
        return None;
    }
    let mut body = Vec::new();
    while let Some(chunk) = resp.chunk().await.ok()? {
        if body.len() + chunk.len() > CDN_MAX_PACKAGE_SIZE as usize {
            return None;
        }
        body.extend_from_slice(&chunk);
    }
    Some(body)
}

fn now_millis() -> u64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as u64
}
